### ISSUE TITLE
Better explain coordinate system in vg.proto

### ISSUE THREAD

**Issue by glennhickey** [vgteam/vg#127]
From existing docs it's not clear what a Position's offset means in the case of a reverse Mapping.  Is it always forward relative or strand relative? In other words, which of the following would be used to represent a mapping over the reverse of a node of length 10 (from right-to-left of node)?  

"mapping": [ {
"is_reverse": true,
          "position": {
            "node_id": 1000,
            "offset": 0
          }]

OR 

"mapping": [ {
"is_reverse": true,
          "position": {
            "node_id": 1000,
            "offset": 9
          }]
          
I think it's the second, but would be nice to confirm, and add details to vg.proto...

**Comment by ekg** [comment]
I think it's the second. I can't remember how this is implemented though.

**Comment by glennhickey** [comment]
Yeah looking like the second.  vg validate will show me the way.

**Comment by ekg** [comment]
Uh oh vg validate isn't that smart yet...

**Comment by glennhickey** [comment]
Doh!

But I've been distracted by another problem that I didn't notice before vg
validate:  Paths with cycles aren't supported.  When importing them,
duplicated mappings are silently deleted, then validate complains about
missing edges in the resulting path.  Though spelled out in a comment in
path.cpp, this took me way too long to find.

// problem, won't allow us to keep multiple identical mapping to the same
node,
// as will happen with looping paths

For some reason, I was under the impression cycle support had been
added.... Do you think this is difficult to add?  Many of our graphs have
paths with cycles...

**Comment by ekg** [comment]
I thought I had fixed this, but I guess not. Sadness.

It should be easy enough to fix. xg supports it.

**Comment by ekg** [comment]
I wonder if this is causing other issues.

**Comment by glennhickey** [comment]
OK, I added an issue for it.

Luckily, we haven't been using vg paths for anything so far (though that
will change), as far as I'm aware.  The only exception was the surjected
bams I was using for the variant comparison, and this issue my be related
to my nonsense results for that.

**Comment by glennhickey** [comment]
No idea, but it's encouraging at least that vg validate picked this up
(however indirectly) right away.

**Comment by ekg** [comment]
There are two ways to represent a reverse alignment. You can represent the path in "forward" orientation and then set the alignment `is_reverse` flag. Or, you can have a path in which the mappings marked with the `is_reverse` flag.

I prefer the second way. It seems more conceptually clear. However, the first is simpler for DAG-like graphs. This needs to be resolved and implemented or there will be a lot of confusion. Suggestions?

**Comment by ekg** [comment]
Right now the validation code determines the edge orientation from the mapping orientations. If we reverse path representations in alignment (flipping all is_reverse flags and setting the alignment.is_reverse flag to indicate that we've done this) then the path no longer contains enough information to describe a valid path. Edges that the validator assumes should exist don't. I ran into this while working on #143.

Problems like this can be avoided by using the is_reverse flag to indicate the state of the path, but not modifying the path. (I wrote "muddifying" and actually that might be a better description of what happens...)

**Comment by ekg** [comment]
xg will not handle paths that don't perfectly match a series of nodes. This isn't ideal, but I'm not sure how to handle it.

**Comment by ekg** [comment]
Presently mappings have starting positions like this:

```
forward:

char pos:       0 1 2 3
character:      A T G C
offset+:       | | | |
offset+ pos:   0 1 2 3

reverse:

char pos:       0 1 2 3
character:      A T G C
offset-:         | | | |
offset- pos:     0 1 2 3
```

I'm not sure that this is the most-intuitive approach.

For instance, would it be easier to think about if the offsets referred to the same thing in the forward and reverse?

**Comment by glennhickey** [comment]
From my end, what I'd like to see is
* reverse strand coord system is clearly documented somehwere, whatever it is
* only one way to represent any given path/alignment (would mean getting rid of Alignment.is_reverse)

For the offset, I'm in favour of the status quo (forward relative) just because I'm somewhat invested in it.  
Changing wouldn't be the end of the world though, so long as it's final.

**Comment by adamnovak** [comment]
I also like the forward-relative notation (so for reverse perfect mappings the offset is the last base on the node). I'm pretty sure that that's how it's implemented in vg right now (when it is implemented). I can successfully vg validate paths with mappings set to is_reverse with offsets set like this, and I can't with offsets set up the other way.

@ekg it doesn't sound in your example like you're updating the offsets. Maybe that confuses validate?

Dropping alignment.is_reverse might make sense? It would mean, though, that you can't ever reverse the sequence of an input read in your alignment output. You would always have to articulate it in the orientation you got it. This could be undesirable if the vg aligner has some notion of which way in the graph is "forward", and the read as read really does go "backward". It's also another difference from SAM/BAM.

**Comment by ekg** [comment]
Here is an example of a perfect reversed mapping against a 20bp node:

```js
{
  "rank": 1,
  "position": {
    "node_id": 4,
    "offset": 19
  },
  "edit": [
    {
      "from_length": 20,
      "to_length": 20
    }
  ],
  "is_reverse": true
}
```

So here the offset 19 means that you are 20bp from the end you're heading to. This seems really hard to reason about because we have to remember that the semantics of the offset change if the is_reverse flag is set.

I prefer to think about mappings against a node's reverse as if that reverse were another node in the "forward" direction. I think this is cleaner than pulling the forward offset over to the reverse strand to describe offsets there.

To motivate this, here are two examples of what I'll implement:

```js
{
  "rank": 1,
  "position": {
    "node_id": 1,
    "offset": 0
  },
  "edit": [
    {
      "from_length": 7,
      "to_length": 7
    }
  ]
}
```

```
seq+:        G A T T A C A
offset+:  → 0 1 2 3 4 5 6
```

Then we have reverse.

```js
{
  "rank": 1,
  "position": {
    "node_id": 1,
    "offset": 0
  },
  "edit": [
    {
      "from_length": 7,
      "to_length": 7
    }
  ],
  "is_reverse": true
}
```

```
seq-:     C T A A T G T
offset-:   6 5 4 3 2 1 0 ←
```

You could also reverse this so the offsets were increasing and it would have the same meaning.

This is already the format used in GCSA2. I'll document the change and then embark on a review to fix the offending cases.

**Comment by ekg** [comment]
As for the alignment.is_reverse flag, I think it should go away and we should rely only on the orientation information in mappings to describe the alignment's path. As these have is_reverse flags then there is no need to keep it at the level of the whole alignment. It's an anachronism from BAM.

Now, I also can't see a way of getting around compatibility with BAM. If we want to invert alignment paths so they can be described relative to the forward strand, we'll either need to do that internally or rewrite the conversion at some other layer. So paths that are reversed have utility.

I'm inclined to do something evil and move is_reverse to paths. If set, it means that the original sequence that the path represents is the reverse complement of that which we'd read out by traversing the path mappings through the graph. In this process of switching this over I can review the use of this kind of reversal and ensure it's only used in surjecting GAM to BAM.

**Comment by pd3** [comment]
What orientation is Edit.sequence with respect to Node.sequence when Mapping.is_reverse flag is set?

**Comment by ekg** [comment]
It should be the reverse complement, as if the edits are describing the transformation of a node that is the reverse complement of the orientation that is stored in the graph.

**Comment by ekg** [comment]
Watch out, I'm moving is_reverse into `Position`!

**Comment by adamnovak** [comment]
This is definitely a solution. I'll have to revise my core graph stuff, but it shouldn't be too hard. Better to change it now than later.

This is going to break backwards compatibility with existing stored alignments and the paths in stored graphs, if they use the flag, right? Or are you going to leave the field in the protobuf, somehow deprecate it, and write conversion code on input?

**Comment by ekg** [comment]
I've mostly implemented a fix. Will try to push this evening. I am fine
with breaking formats. If we need to work with older files we have releases
available. I am incrementing protobuf field ids properly so the old files
will still be partly readable.

**Comment by adamnovak** [comment]
This problem is back and super confusing! I'm going to standardize by fiat on what "vg map" is doing in practice, and change everything else to work like that.

What "vg map" does is give the offset relative to the left end of the oriented view of the node.

Here's an example. I grabbed the middle of node 77, leaving 3 bases on the left and 5 bases on the right, and aligned it:

```
vg construct -r test/small/x.fa -v test/small/x.vcf.gz | vg map -V -s GCTGTGAAGATTAAATTAGGTGAT -J -k 16 - | jq .

{
  "path": {
    "mapping": [
      {
        "position": {
          "node_id": 77,
          "offset": 3
        },
        "edit": [
          {
            "from_length": 24,
            "to_length": 24
          }
        ],
        "rank": 1
      }
    ]
  },
  "sequence": "GCTGTGAAGATTAAATTAGGTGAT",
  "mapping_quality": 141,
  "score": 24,
  "identity": 1
}
```

Note that the offset is 3 (and counting from the start of the node in its true orientation).

Now align the reverse complement of that sequence:

```
vg construct -r test/small/x.fa -v test/small/x.vcf.gz | vg map -V -s ATCACCTAATTTAATCTTCACAGC -J -k 16 - | jq .

{
  "sequence": "ATCACCTAATTTAATCTTCACAGC",
  "path": {
    "mapping": [
      {
        "position": {
          "node_id": 77,
          "offset": 5,
          "is_reverse": true
        },
        "edit": [
          {
            "to_length": 24,
            "from_length": 24
          }
        ],
        "rank": 1
      }
    ]
  },
  "mapping_quality": 141,
  "identity": 1,
  "score": 24
}
```

Note that we now have an offset of 5, and `is_reverse` is true: we're counting from the end of the node in its true orientation, or the start of the node in its local, aligned-to orientation.

Basically, offset counts the number of unused bases before the mapping starts, in the local orientation of the node, reading from its locally-oriented beginning.

This is consistent with the textual description in `vg.proto`, but the diagrams said something different. I'm going to make the diagrams look like this instead:

```
// seq+        G A T T A C A
// offset+  → 0 1 2 3 4 5 6 7
//
// seq-        C T A A T G T
// offset+  → 0 1 2 3 4 5 6 7
//
// Or both at once:
//
// offset-    7 6 5 4 3 2 1 0 ←
// seq+        G A T T A C A
// offset+  → 0 1 2 3 4 5 6 7
```
